src/DataStructures/LinkedMessageQueue.hpp

```diff
@@ -0,0 +1,178 @@
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) noexcept {
+    p | id;
+    p | previous;
+  }
+};
+
+template <typename Id>
+bool operator==(const LinkedMessageId<Id>& a,
+                const LinkedMessageId<Id>& b) noexcept {
+  return a.id == b.id and a.previous == b.previous;
+}
+
+template <typename Id>
+bool operator!=(const LinkedMessageId<Id>& a,
+                const LinkedMessageId<Id>& b) noexcept {
+  return not(a == b);
+}
+
+template <typename Id>
+std::ostream& operator<<(std::ostream& s,
+                         const LinkedMessageId<Id>& id) noexcept {
+  return s << id.id << " (" << id.previous << ")";
+}
+
+namespace std {
+template <typename Id>
+struct hash<LinkedMessageId<Id>> {
+  size_t operator()(const LinkedMessageId<Id>& id) const noexcept {
+    // For normal use, any particular .id should only appear with one
+    // .previous, and vice versa, so hashing only one is fine.
+    return std::hash<Id>{}(id.id);
+  }
+};
+}  // namespace std
+
+/// \cond
+template <typename Id, typename QueueTags>
+class LinkedMessageQueue;
+/// \endcond
+
+/// \ingroup DataStructuresGroup
+/// A collection of queues of asynchronously received data
+///
+/// This class is designed to collect asynchronously received
+/// messages, possibly from multiple sources, until sufficient data
+/// has been received for processing.  Each tag in \p QueueTags
+/// defines a queue of messages of type `Tag::type`.
+template <typename Id, typename... QueueTags>
+class LinkedMessageQueue<Id, tmpl::list<QueueTags...>> {
+  static_assert(sizeof...(QueueTags) > 0,
+                "Must have at least one message source.");
+
+ public:
+  using IdType = Id;
+
+  /// Insert data into a given queue at a given ID.  All queues must
+  /// receive data with the same collection of \p id_and_previous, but
+  /// are not required to receive them in the same order.
+  template <typename Tag>
+  void insert(const LinkedMessageId<Id>& id_and_previous,
+              typename Tag::type message) noexcept;
+
+  /// The next ID in the received sequence, if all queues have
+  /// received messages at that ID.
+  std::optional<Id> ready_id() const noexcept;
+
+  /// All the messages received at the time indicated by `ready_id()`.
+  /// It is an error to call this function if `ready_id()` would
+  /// return an empty optional.  This function removes the stored
+  /// messages from the queues, advancing the internal sequence ID.
+  tuples::TaggedTuple<QueueTags...> extract() noexcept;
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) noexcept {
+    p | previous_id_;
+    p | messages_;
+  }
+
+ private:
+  template <typename Tag>
+  struct Optional {
+    using type = std::optional<typename Tag::type>;
+  };
+
+  using OptionalTuple = tuples::TaggedTuple<Optional<QueueTags>...>;
+
+  std::optional<Id> previous_id_{};
+  std::unordered_map<std::optional<Id>, std::pair<Id, OptionalTuple>>
+      messages_{};
```

Comment:
Why are the messages stored corresponding to the previous id and not the current id? I would have though messages correspond to a current id.